visual-js/visual-playwright/src/api.ts

```diff
@@ -0,0 +1,325 @@
+ */
+export const sauceVisualCheck = async (
+  page: Page,
+  testInfo: TestInfo,
+  name: string,
+  options?: Partial<SauceVisualParams>,
+) =>
+  await takePlaywrightScreenshot(
+    page,
+    {
+      deviceName: testInfo.project.name,
+      testName: testInfo.title,
+      suiteName: testInfo.titlePath.slice(0, -1).join('/'),
+    },
+    name,
+    options,
+  );
+
+export const takePlaywrightScreenshot = async (
+  page: Page,
+  info: {
+    testName: string | undefined;
+    suiteName: string | undefined;
+    deviceName: string | undefined;
+  },
+  name: string,
+  options?: Partial<SauceVisualParams>,
+) => {
+  const { testName, suiteName, deviceName } = info;
+  const { buildId } = getOpts();
+
+  if (!buildId) {
+    return;
+  }
+
+  const browser = page.context().browser();
+  const {
+    screenshotOptions = {},
+    clipSelector,
+    delay,
+    captureDom,
+    ignoreRegions: userIgnoreRegions,
+    diffingMethod,
+  } = options ?? {};
+  const { animations = 'disabled', caret } = screenshotOptions;
+  let ignoreRegions: RegionIn[] = [];
+
+  const promises: Promise<unknown>[] = [
+    // Wait for all fonts to be loaded & ready
+    page.evaluate(() => document.fonts.ready),
+  ];
+
+  if (delay) {
+    // If a delay has been configured by the user, append it to our promises
+    promises.push(new Promise((resolve) => setTimeout(resolve, delay)));
+  }
+
+  if (userIgnoreRegions) {
+    promises.push(
+      (async (): Promise<void> => {
+        const filterIgnoreRegion = (
+          region: RegionIn | string,
+        ): region is RegionIn => typeof region !== 'string';
+        const filterIgnoreSelector = (
+          region: RegionIn | string,
+        ): region is string => typeof region === 'string';
+
+        const selectors = userIgnoreRegions.filter(filterIgnoreSelector);
+        let selectorRegions: RegionIn[] = [];
+
+        if (selectors.length) {
+          selectorRegions = await page.evaluate(
+            ({ selectors }) => {
+              const selectorRegions: RegionIn[] = [];
+              selectors.forEach((selector) => {
+                const elements = document.querySelectorAll(selector);
+                elements.forEach((element) => {
+                  const rect = element.getBoundingClientRect();
+
+                  selectorRegions.push({
+                    name: selector,
+                    x: Math.round(rect.x),
+                    y: Math.round(rect.y),
+                    height: Math.round(rect.height),
+                    width: Math.round(rect.width),
+                  });
+                });
+              });
+              return selectorRegions;
+            },
+            { selectors },
+          );
+        }
+
+        ignoreRegions = [
+          ...userIgnoreRegions.filter(filterIgnoreRegion),
+          ...selectorRegions,
+        ].filter((region) => 0 < region.width * region.height);
```

Comment:
We should probably check that both width / height are positive.
Having 2 negative value will result in the region to pass that filter while it's invalid. Isn't it ?

Reply:
Added a `Math.max` here for both.